rinstall.1
.Dd October 8, 2026
.Dt RINSTALL 1
.Os
.Sh NAME
.Nm rinstall
.Nd fetch and install scripts from local or http source
.Sh SYNOPSIS
.Nm rinstall
.Op Fl a Ar location
.Op Fl m Ar mode
.Op Fl o Ar owner:group
.Ar source
.Op Ar target
.Sh DESCRIPTION
.Nm
is shipped to remote machines by
.Xr rset 1
to provide a standard method of installing files made available over HTTP, or
from the current working directory for files already staged.
.Pp
The
.Ar target
must be an absolute path, and is created if it does not exist.
If the
.Ar target
exists but is not the same as
.Ar source ,
then
.Xr diff 1
is used to display the difference before it is updated.
If the
.Ar target
is omitted, the
.Ar source
is placed in the staging directory defined by the
.Ev SD
environment variable.
.Pp
If the source is defined absolute path it must exist on a file system on the
remote host.
If the source is defined with a relative path, then
.Nm
first checks if the source already exists in the
.Ev SD
and then tries to fetch it from the HTTP resource defined by
.Ev INSTALL_URL
or the
.Fl a
option.
.Pp
.Nm
always checks for a local copy of the
.Ar source
before attempting to fetch a file from the location defined by the environment
variable
.Ev INSTALL_URL .
.Pp
The arguments are as follows:
.Bl -tag -width Ds
.It Fl a
URL to an alternative file location if source was not found using
.Ev INSTALL_URL .
.It Fl m Ar mode
Set file mode.
This argument is passed to
.Xr chmod 1 .
.It Fl o Ar owner:group
Set owner and/or group.
This argument is passed to
.Xr chown 8 .
.El
.Sh ENVIRONMENT
.Bl -tag -width Ds
.It Ev RINSTALL_DIFF_ARGS
Arguments for
.Xr diff 1
tool.
The default is
.Qq -U 2 .
.El
.Sh EXIT STATUS
The
.Nm
utility exits with one of the following values:
.Pp
.Bl -tag -offset indent -compact -width Ds
.It 0
Target file was installed or updated
.It 1
No changes were made
.It 2
Unknown Operating System
.It 3
Unable to fetch requested file
.El
.Sh EXAMPLES
Install configuration file and set permissions
.Pp
.Dl $SD/rinstall -m 644 -o www:www resolv.conf /var/www/etc/
.Pp
Fetch latest copy of sources and extract
.Pp
.Dl $SD/rinstall wodpress.tar.gz
.Dl tar -xzf $SD/wordpress.tar.gz -C /var/www
.Pp
Fetch a file from an alternative location and store it in the staging directory
under tools/sys/ if it's not already there and it's not available at
$INSTALL_URL/tools/sys/sometool.tgz
.Pp
.Dl $SD/rinstall -a https://example.org/files/tool.tar.gz tools/sys/sometool.tgz
.Sh HISTORY
The
.Nm
utility was introduced in rset 0.1


NEWS
= Release History

== 3.5: October 8, 2026

 - Dynamically allocate route labels; use arbitrary limit of 10000 routes
 - Set a limit of 1000 for range expansion
 - New 'init' command with example configurations for 'alpine' and 'openbsd'
 - New 'wenv' command, use custom log format to supersede '-e' flag
 - Display multibyte characters in error messages
 - Enable environment_file= to accept multiple pathnames
 - pln(5): no longer document single-use options
 - pln(5): rewrite man page and organize in three major sections
 - renv(1): require closing quote, reject trailing whitespace

== 3.4: August 3, 2026

 - Match route labels using regular expressions, or any alias using exact hostnames
 - Do not count zero-width regex as a match
 - All utilities: add usage hint and '-h' flag to display option summary
 - Detect string copy overflow and abort with an error
 - rinstall(1) create directories using the same owner as ${SD}
 - miniquark(1) compare HTTP If-Modified-Since times using GMT
 - miniquark(1) adjust options to resemble nc(1)
 - Use 'ustar' format for transfering files on OpenBSD
 - Replace 'tmppath' pledge with 'wpath cpath' on OpenBSD
 - Exit without cleanup if LOCAL_CONTROL_SOCKET already exists
 - Wait 0.5 seconds for SSH control socket to clear

== 3.3: January 17, 2026

 - Use session_id for remote stage dir name
 - Access miniquark logs using HTTP_TRACE
 - Access ssh invocation using SSH_TRACE
 - Do not set INSTALL_URL, rely on default in rinstall(1)
 - Exit on failure for malloc(2), pipe(2), socket(2), bind(2), strdup(2)
 - Allow export path on host labels used for archive/restore
 - pln(5) Ensure export paths are only specified on route labels
 - pln(5) Raise error for indented text in unexpected context
 - miniquark(1) listen on multiple addresses, log each bind address
 - miniquark(1) remove chroot feature
 - rinstall(1) default to curl

== 3.2: February 6, 2025

 - Update MANPREFIX for FreeBSD
 - Format source using clang-format and rubocop
 - rsub(1) allow diff(1) behavior to be modified using RSUB_DIFF_ARGS
 - rsub(1) refactor based on patterns used in rinstall(1)
 - rsub(1) remove temporary source file
 - Ensure top-level directory permissions are not world-readable
 - Avoid leaving a NUL character at the end of local execution script
 - Use native strtonum(3) on MacOS

== 3.1: September 24, 2024

 - Add begin= and end= options
 - Add archive and restore options
 - Allow IPv6 addresses to be used in routes
 - Add out-of-bounds check to options
 - Use 'set -a' instead of 'sh -a' to avoid environment contamination
 - Enable parallel execution using background workers
 - rinstall(1) print name of missing source with absolute path
 - rinstall(1) adjust comments and error messages

== 3.0: February 26, 2024

 - Check for ssh-agent before updating _rutils
 - Correct use of getopt(3) to allow compact flags
 - Use RSET_HOST_CONNECT_ERROR to format failed connection or stagedir initialization
 - Use RSET_LABEL_EXEC_ERROR if ssh(1) or sh(1) terminates during script execution
 - Add log format variable '%s' for generated session ID
 - Remove '-l' option
 - Tests: use curl(1) on Linux
 - Detect and use native strlcpy(3), arc4random(3) on Linux
 - Upload _rutils and exported paths in two steps for compatibility with busybox tar(1)
 - rsub(1): escape ampersands found in '-l' argument
 - rinstall(1) allow diff(1) behavior to be modified using RINSTALL_DIFF_ARGS
 - rinstall(1) add '-a' option for defining an alternate source location
 - rinstall(1) saving a file in $SD no longer depends the current directory
 - rinstall(1) failure running diff(1) no longer reports "files are different"

== 2.9: December 22, 2023

 - Add flag '-E' for overriding remote environment variables
 - rset(1)/renv(1) extended to store session-level state using name="value"
   arguments
 - renv(1) improved error handling, one variable per line
 - Immediately raise errors and abort if environment or environment_file options
   contain syntax errors
 - rinstall(1) downloads sources to the staging directory without a temporary
   file to avoid duplication
 - rinstall(1) prioritize curl with fallback to wget on Linux
 - rinstall(1) recreates file path when fetching over HTTP
 - rinstall(1) allows empty files to be installed
 - Only support '-l' flag with dry-run
 - New custom log format for each stage of execution defined by environment
   variables

== 2.8: November 13, 2023

 - No longer set ROUTE_LABEL, and LABEL on remote host
 - New renv(1) utility for validating environment variables
 - New environment_file option for defining variables
 - New environment option for defining variables inline
 - Improve parsing by prioritizing parameters before labels
 - '-l' requires an option value

== 2.7: October 28, 2023

 - Avoid opening temporary files twice
 - Connect to the hostname specified, not the primary label name
 - Do not start web server during dry run
 - Execute all hostnames specified on the command line
 - Support for hostname sequence ranges in routes file

== 2.6: September 6, 2023

 - rinstall: use wget with fallback to curl except for *BSD
 - configure: use TARGET_OS to override the output of uname(1)
 - MacOS: use tar -c --no-xattrs
 - Use mkstemp(3) to avoid compiler warnings
 - Add bundler/inline and install webrick locally to run tests

== 2.5: December 16, 2022

 - rinstall: ignore environment variable http_proxy
 - rinstall: Use shell getopts to parse arguments
 - Raise error if a label alias begins with a space

== 2.4: May 3, 2022

 - Formatting improvements to tests
 - Route labels may be a comma-separated list

== 2.3: November 17, 2021

 - Tweaks to unveil(2)
 - rinstall(1) accepts a directory instead of a target, similar to cp(1)

== 2.2: June 16, 2021

 - Raise an error if the output of local execution does not end with '\n'
 - Rename "Remote Sequential Execution Tool" to "Remote Staging Execution Tool"
 - Fix miniquark arg parsing on arm

== 2.1: April 3, 2021


rinstall.sh
#!/bin/sh
# rinstall
# A helper utility for rset(1)
# Install files from the local staging area or a remote URL

set_defaults() {
	ret=1           # global exit status
	fetched=0       # set to 1 when source was fetched via HTTP
	samedir=0       # set to 1 when target is $SD (Staging Directory)
	source_local=0  # set to 1 if source is defined with an absolute path
	owner=""
	mode=""
	alt_location=""
	: ${INSTALL_URL:=http://localhost:6000}
	: ${RINSTALL_DIFF_ARGS:="-U 2"}
}

main() {
	set_defaults
	parse_args "$@"
	init
	set_source_target_vars "$arg_src" "$arg_dst"

	# If source does not exist then it was not found on a local file system
	# (with an absolute path) or in the current directory (normally $SD)
	if [ ! -f "$source" ]; then
		if [ $source_local -eq 1 ]; then
			>&2 echo "rinstall: source $source with absolute path does not exist"
			exit 1
		fi
		# If file has not been staged, fetch over HTTP
		# Fail if download fails
		[ -f "$SD/$source" ] || download_source
	fi

	# Source file exists: check the difference between source and target
	check_diff_source_target

	# Install a file only if source and target are different
	install_target

	# Target file exists: try changing owner and/or permissions but do not fail
	set_mode_owner

	exit $ret
}

usage() {
	>&2 echo "release: ${release}"
	>&2 echo "usage: rinstall [-a location] [-m mode] [-o owner:group] source [target]"
	if [ -z "$1" ]; then
		echo >&2 "hint: use -h to display option summary"
		exit 1
	fi

	cat <<- HELP
		summary:
		    -a location      URL to use if source is not found locally
		    -m mode          Arguments passed to chmod(1)
		    -o owner:group   Arguments passed to chown(8)
		docs:
		    man rinstall
	HELP
	exit 1
}

init() {
	if [ -z "$SD" ]; then
		>&2 echo "rinstall: staging directory \$SD is not defined"
		exit 1
	fi
	if ! check_absolute_path "$SD"; then
		>&2 echo "rinstall: $SD is not an absolute path"
		exit 1
	fi

	unset http_proxy
	trap '' HUP
}

parse_args() {
	[ "x$1" = "x-h" ] && usage $1
	while getopts m:o:a: arg; do
		case "$arg" in
			o) owner="$OPTARG" ;;
			m) mode="$OPTARG" ;;
			a) alt_location="$OPTARG" ;;
			?) usage ;;
		esac
	done
	shift $(($OPTIND - 1))
	[ $# -eq 1 -o $# -eq 2 ] || usage
	arg_src="$1"
	arg_dst="$2"
}

set_source_target_vars() {
	# Source can be a local file defined by absolute path or remote defined
	# using a relative path
	source="$1"
	src_path="$(dirname $source)"

	if [ -z "$2" ]; then
		# Target is not defined, try source located in $SD

		# Ensure the source has a relative path by removing leading slashes
		src_rel_path="$(echo "$src_path" | sed 's|^/*||')"

		# prepare a relative path for the target in $SD
		[ -d "$SD/$src_rel_path" ] || mkdir -p "$SD/$src_rel_path" || {
			>&2 echo "rinstall: could not create a relative path: $SD/$src_rel_path"
			exit 1
		}
		fix_permissions

		if [ "$src_rel_path" = "." ]; then
			target="$SD/$(basename "$source")"
		else
			target="$SD/$src_rel_path/$(basename "$source")"
		fi
		samedir=1
	elif [ -d "$2" ]; then
		target="$2/$(basename "$source")"
	else
		target="$2"
	fi

	if ! check_absolute_path "$target"; then
		>&2 echo "rinstall: $target is not an absolute path"
		exit 1
	fi

	if check_absolute_path "$src_path"; then
		source_local=1
	fi
}

download_source() {
	# Source file does not exist on a local file system, thus it is remote
	# and it should not be defined with an absolute path
	if check_absolute_path "$src_path"; then
		>&2 echo "rinstall: absolute path is not allowed for HTTP fetch: $source"
		exit 1
	fi

	# Reconstruct the source's relative path
	if [ ! -d "$SD/$src_path" ]; then
		mkdir -p "$SD/$src_path" || {
			>&2 echo "rinstall: could not create a relative path: $SD/$src_path"
			exit 1
		}
		fix_permissions
	fi

	fetch_file "$INSTALL_URL/$source"
	if [ $? -ne 0 ]; then
		if [ -n "$alt_location" ]; then
			echo "rinstall: using alternate source: $alt_location"
			fetch_file "$alt_location"
			[ $? -eq 0 ] || {
				>&2 echo "rinstall: unable to fetch $alt_location"
				exit 3
			}
		else
			>&2 echo "rinstall: unable to fetch $INSTALL_URL/$source"
			exit 3
		fi
	fi
	fetched=1
}

fetch_file() {
	case $(uname) in
		OpenBSD)
			ftp -o "$SD/$source" -n "$1"
			;;
		FreeBSD)
			fetch -qo "$SD/$source" "$1"
			;;
		*)
			if command -v curl > /dev/null; then
				curl -fsLo "$SD/$source" "$1"
			else
				wget -qO "$SD/$source" "$1"
			fi
			;;
	esac
	return $?
}

check_diff_source_target() {
	# Set the 'create' flag:
	#   0 - files are the same
	#   1 - a new file
	#   2 - files are different

	# If source was fetched, then the source is located in $SD, otherwise the
	# source is on a local file system
	if [ $fetched -eq 1 -o $source_local -eq 0 ]; then
		src_file="$SD/$source"
	else
		src_file="$source"
	fi

	if [ -e "$target" ]; then
		if [ $samedir -eq 1 ]; then
			create=0
			[ $fetched -eq 0 ] || {
				ret=0
				echo "rinstall: fetched $target"
			}
		else
			if output="$(diff $RINSTALL_DIFF_ARGS "$target" "$src_file" 2>&1)"
			then
				create=0
			else
				case $? in
					1)  create=2
						[ -z "$output" ] || echo "$output"
						;;
					*)  >&2 echo "rinstall: $output"
						exit 1
						;;
				esac
			fi
		fi
	else
		create=1
	fi
}

install_target() {
	if [ $create -ne 0 ]; then
		cp "$src_file" "$target" && ret=0 || {
			>&2 echo "rinstall: could not copy $src_file into $target"
			exit 1
		}
		if [ $create -eq 1 ]; then
			echo "rinstall: created $target"
		fi
	fi
}

set_mode_owner() {
	if [ -n "$owner" ]; then
		chown $owner "$target"
	fi
	if [ -n "$mode" ]; then
		chmod $mode "$target"
	fi
}

check_absolute_path() {
	case $(dirname "$1") in
		/*) return 0
			;;
		*)  return 1
			;;
	esac
}

fix_permissions() {
	uid=$(stat -f '%u' $SD 2> /dev/null) || uid=$(stat -c '%u' $SD)
	find $SD -type d -not -user $uid -exec chown $uid {} ';'
}

main "$@"
